crates/rpc/rpc/src/eth/api/server.rs

```diff
@@ -199,14 +199,26 @@ where
             return Ok(FeeHistory::default())
         }
 
-        let Some(end_block) = self.inner.client.block_number_for_id(newest_block).to_rpc_result()? else { return Err(EthApiError::UnknownBlockNumber.into())};
+        let head_block = self.inner.client.chain_info().to_rpc_result()?.best_number;
+
+        let Some(end_block) = self.inner.client.block_number_for_id(newest_block).to_rpc_result()? else { return Err(EthApiError::RequestedBlockBeyondHead(head_block).into())};
 
         if end_block < block_count {
-            return Err(EthApiError::InvalidBlockRange.into())
+            return Err(EthApiError::RequestedBlockBeyondHead(head_block).into())
         }
 
         let start_block = end_block - block_count;
 
+        // Query canonical block hashes for the whole block range to compare them against cache
+        // We need it due to the possibility of reorgs when block numbers match but hashes don't
+        let block_hashes =
+            self.inner.client.canonical_hashes_range(start_block..=end_block).to_rpc_result()?;
```

Comment:
this is a lot of overhead I think?

this makes me wonder if we instead should populate the cache with every new block instead? this way reorged blocks are simply replaced?

Reply:
yeahhhh, agree. If my understanding is right, we will still need this part with canonical hashes retrieval because even if we had cache populated with every new block, for older historical blocks (`latest-1024` and older) we will need to populate cache with new blocks and respective caches.

Reply:
regarding the cache population on every new block: I think it can be implemented with the existing `StatusUpdater` trait that we currently use only for P2P updates. It can be extended to RPC-related updates with something like `MultiStatusUpdater` which send new updates to every updater in `Vec<Box<dyn StatusUpdater>>`. Also, we'll need to spawn a listener inside the `EthApi` that will receive the new head updates from its `StatusUpdater`.

Reply:
also I started wondering how (and if) geth handles reorgs in the fee history oracle, and couldn't find any evidence of it. The `cacheKey` struct [contains](https://github.com/ethereum/go-ethereum/blob/b40c10916c13660472d3071f7e5c1fcdee9aceec/eth/gasprice/feehistory.go#L59-L62) only `number` and `percentiles`, and on successful cache retrieval [there's no check](https://github.com/ethereum/go-ethereum/blob/b40c10916c13660472d3071f7e5c1fcdee9aceec/eth/gasprice/feehistory.go#L280-L282) for block hash.

Reply:
we also have this now https://github.com/paradigmxyz/reth/blob/1edb9d4ccdc5e0b4801f8deb77e1f4a0bc19c8d4/crates/interfaces/src/events.rs#L9
which we'll use in RPC. we haven't exactly figured out how this will be used combined with the StatusUpdater exactly, but this event type is supposed to be called for every new block.

Reply:
> or older historical blocks (latest-1024 and older)

is this actually supported?

but yes, having this makes sense.

Reply:
> Also, we'll need to spawn a listener inside the EthApi that will receive the new head updates from its StatusUpdater.

right, we need a task that listens for new blocks (via event subscription) and inserts into the cache.

Reply:
> is this actually supported?

yep
```console
➜  ~ cast rpc eth_feeHistory $(cast --to-hex 10) $(cast --to-hex 12965005) '[]' | jq
{
  "baseFeePerGas": [
    "0x0",
    "0x0",
    "0x0",
    "0x0",
    "0x3b9aca00",
    "0x430da58e",
    "0x4b6d419d",
    "0x54d7aa96",
    "0x5f714b18",
    "0x6b5c0aa8",
    "0x78c58a1e"
  ],
  "gasUsedRatio": [
    0.9988511759145364,
    0.999269045039813,
    0.999064899888636,
    0.9998319941325032,
    0.9998712916081928,
    0.9995115300443024,
    0.999327338616906,
    0.9997548197409084,
    0.9994411941322008,
    0.9997078251297092
  ],
  "oldestBlock": "0xc5d484"
}
```

Reply:
ok, added the continuous cache updating through `EthApi::start_fee_history_cache`, let me know if there's a better way to do it.